### PR TITLE
Add some hooks and elrondei functions

### PR DESCRIPTION
**Pull request by andreibancioiu**
Hooks:

```
 + GetUserAccount
 + GetShardOfAddress
 + IsSmartContract
 - AccountExists
 - GetBalance
 - GetNonce
 - IsCodeEmpty
 - GetCode 
```

Added missing implementations in blockchain gateway (`GetAllState`).

`elrondei`:

```
 + getOwnerAddress
 + getShardOfAddress
 + isSmartContract
 getOwner -> getSCAddress (partially)
```

### CODE REVIEW

File: arwen/contexts/blockchain.go [hunk 1]
```diff
@@ -41,8 +41,9 @@ func (context *blockchainContext) NewAddress(creatorAddress []byte) ([]byte, err
 }
 
 func (context *blockchainContext) AccountExists(address []byte) bool {
-	exists, _ := context.blockChainHook.AccountExists(address)
-	return exists
+	// Question for review: should we handle the return error perhaps, like: if err == vmcommon.AccountNotFound?
+	account, _ := context.blockChainHook.GetUserAccount(address)
+	return account != nil
```
Comment (sasurobert) [inline arwen/contexts/blockchain.go:46]: check with arwen.ifNil

File: arwen/elrondapi/elrondei.go [hunk 2]
```diff
@@ -69,7 +72,23 @@ func ElrondEIImports() (*wasmer.Imports, error) {
 	imports := wasmer.NewImports()
 	imports = imports.Namespace("env")
 
-	imports, err := imports.Append("getOwner", getOwner, C.getOwner)
+	// TODO: rename getOwner -> getSCAddress when all affected contracts (delegation etc.) are ready
+	imports, err := imports.Append("getOwner", getSCAddress, C.getSCAddress)
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:76]: change it rigth now, contracts will follow this PR :)

File: arwen/elrondapi/elrondei.go [hunk 3]
```diff
@@ -300,10 +319,70 @@ func getOwner(context unsafe.Pointer, resultOffset int32) {
...
 
+//export getOwnerAddress
+func getOwnerAddress(context unsafe.Pointer, resultOffset int32) {
+	blockchain := arwen.GetBlockchainContext(context)
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	owner, err := blockchain.GetOwnerAddress()
+	// Question for review: is this error handling ok?
+	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
```
Comment (sasurobert) [inline arwen/elrondapi/elrondei.go:334]: error handling is ok. although this should not happen.

File: ipc/arwenpart/blockchainGateway.go [hunk 4]
```diff
@@ -357,6 +287,67 @@ func (blockchain *BlockchainHookGateway) GetBuiltinFunctionNames() vmcommon.Func
...
+	}
+
+	response := rawResponse.(*common.MessageBlockchainGetAllStateResponse)
+	return response.AllState, request.GetError()
+}
+
+// GetUserAccount forwards a message to the actual hook
+// TODO: Perhaps cache GetUserAccount()? Since when it is called with address == contract address, the whole code is fetched.
+func (blockchain *BlockchainHookGateway) GetUserAccount(address []byte) (vmcommon.UserAccountHandler, error) {
+	request := common.NewMessageBlockchainGetUserAccountRequest(address)
```
Comment (sasurobert) [inline ipc/arwenpart/blockchainGateway.go:307]: you can create this cache at the level of arwen context. and clean every time

File: mock/account.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
+package mock
+
+import (
+	"math/big"
+)
+
+// Account holds the account info
+type Account struct {
```
Comment (sasurobert) [inline mock/account.go:8]: you already defined this in ipc/common.go

File: ipc/common/account.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	Balance         *big.Int
+	CodeHash        []byte
+	RootHash        []byte
+	Address         []byte
+	DeveloperReward *big.Int
+	OwnerAddress    []byte
+	UserName        []byte
+	CodeMetadata    []byte
+	// Question for review: will this be populated (always?) by the blockchain hook in elrond-go?
+	Code []byte
```
Comment (sasurobert) [inline ipc/common/account.go:19]: only if there is code.

Comment (sasurobert) [reply ipc/common/account.go:19]: thus only if it is a smart contract

File: arwen/contexts/blockchain.go [hunk 7]
```diff
@@ -41,8 +41,9 @@ func (context *blockchainContext) NewAddress(creatorAddress []byte) ([]byte, err
 }
 
 func (context *blockchainContext) AccountExists(address []byte) bool {
-	exists, _ := context.blockChainHook.AccountExists(address)
-	return exists
+	// Question for review: should we handle the return error perhaps, like: if err == vmcommon.AccountNotFound?
+	account, _ := context.blockChainHook.GetUserAccount(address)
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply arwen/contexts/blockchain.go:45]: Fixed.

File: arwen/contexts/blockchain.go (hunk 1 above)
Comment (andreibancioiu) [reply arwen/contexts/blockchain.go:46]: Fixed.

File: mock/account.go (hunk 5 above)
Comment (andreibancioiu) [reply mock/account.go:8]: The current mocks depends on this - and the struct has some mock-related features. Can we keep it like this? Fixed by renaming to `AccountMock`.

File: arwen/elrondapi/elrondei.go (hunk 2 above)
Comment (andreibancioiu) [reply arwen/elrondapi/elrondei.go:76]: Fixed.

File: ipc/arwenpart/blockchainGateway.go (hunk 4 above)
Comment (andreibancioiu) [reply ipc/arwenpart/blockchainGateway.go:307]: We have some older todo's:
```
// TODO must get the code from the OutputAccount, if present
```
These will solve this issue as well. Will be done in another PR - is it ok?

File: arwen/contexts/blockchain.go [hunk 8]
```diff
@@ -108,18 +117,24 @@ func (context *blockchainContext) GetCodeHash(addr []byte) ([]byte, error) {
 	return context.host.Crypto().Keccak256(code)
 }
 
-func (context *blockchainContext) GetCode(addr []byte) ([]byte, error) {
+func (context *blockchainContext) GetCode(address []byte) ([]byte, error) {
 	// TODO must get the code from the OutputAccount, if present
```
Comment (sasurobert) [inline arwen/contexts/blockchain.go:121]: delete toDO

File: arwen/elrondapi/elrondei.go [hunk 9]
```diff
@@ -300,10 +319,69 @@ func getOwner(context unsafe.Pointer, resultOffset int32) {
...
+		return 0
+	}
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.IsSmartContract
+	metering.UseGas(gasToUse)
+
+	isSmartContract := blockchain.IsSmartContract(address)
+	if isSmartContract {
+		return 1
```
Comment (ccorcoveanu) [inline arwen/elrondapi/elrondei.go:379]: Maybe we should have constants for TRUE and FALSE?

File: arwen/contexts/blockchain.go (hunk 8 above)
Comment (andreibancioiu) [reply arwen/contexts/blockchain.go:121]: Deleted, and added Jira task: https://elrondnetwork.atlassian.net/browse/EN-6829

File: arwen/elrondapi/elrondei.go (hunk 9 above)
Comment (andreibancioiu) [reply arwen/elrondapi/elrondei.go:379]: Fixed by adding a helper function, `BooleanToInt`. Is this ok?

**Pull request merged**